Add release process (#97)
kubearchive/kubearchive: docs/modules/ROOT/pages/contributors/release.adoc

= Release Process

> This document explains things related to the KubeArchive release.

Reviewer: I suggest this:

```suggestion
This document explains the requirements for and the process of a KubeArchive release.
```

== Commits

The KubeArchive release process scans commits from the latest tag up to the current
commit. The commits must mention the GitHub Pull Requests they come from on their
title (first line). A commit merged from the Pull Request with ID 100 should have
this title:

[source,text]
----
Add feature A (#100)
----

The Pull Request mention is added by default when the commits are squashed on
Pull Request merge.

== Pull Request Labels

KubeArchive follows the
link:https://semver.org/[Semantic Versioning]
system with a leading `v`: `vMAJOR.MINOR.PATCH`. To summarize, `MAJOR` releases
break user-facing interfaces, `MINOR` releases introduce new features and `PATCH`

> versions introduce bugfixes or documentation changes.

Reviewer: I think "bug fixes" should be two words.

```suggestion
versions introduce bug fixes or documentation changes.
```

> To decide the next version the KubeArchive release process relies on the labels
> on the GitHub Pull Requests since the latest release. The following labels impact
> which number is assigned to the next release:

Reviewer: "To decide" feels wrong here, and I think this first sentence would be better in the active sense.

```suggestion
The KubeArchive release process relies on the labels on the GitHub Pull Requests
when determining the next version number for the release. The following labels
impact which number is updated in the next release:
```

* `kind/breaking`: next release is a MAJOR version. MINOR and PATCH numbers are reset to 0.
* `kind/feature`: if `kind/breaking` is not present, next release is a MINOR version.
PATCH number is reset to 0.
* `kind/documentation` and `kind/bug`: if `kind/breaking` and `kind/feature` are
not present, next release is a PATCH version.

> Pull Request require at least one of these labels.

Reviewer: It should be either "A Pull Request requires..." or "Pull Requests require..."

```suggestion
A Pull Request requires at least one of these labels.
```

== Release Notes

The KubeArchive release process creates release notes for users. These notes are
generated from the GitHub Pull Requests that contain a code block annotated with
`release-note`:

[source,text]
----
```release-note
Introduced new feature called `feature-a`. To enable it, set to `true` the
`feature-a` flag in the `feature-flags` ConfigMap in the `kubearchive` namespace.
```
----

> To skip the Pull Request for release notes the block must contain the string `NONE`:

Reviewer: Suggested rewording:

```suggestion
To not include release notes from the Pull Request the release-note block must contain the string `NONE`:
```

[source,text]
----
```release-note
NONE
```
----

== Steps

. Merge the changes needed into `main`.
. Make sure the latest commit in `main` has passed every test.
. Go to the Actions tab on the KubeArchive repository.
. Trigger the "Release" workflow.

> == Side-effects

Reviewer: I would not call these "side effects". That usually implies unexpected or unwanted behavior. I suggest something like "Release Results" or "Results of the Release Workflow".

The KubeArchive release process relies on the labels on the GitHub Pull Requests
when determining the next version number for the release. The following labels
impact which number is updated in the next release:

* `kind/breaking`: next release is a MAJOR version. MINOR and PATCH numbers are reset to 0.
* `kind/feature`: if `kind/breaking` is not present, next release is a MINOR version.
PATCH number is reset to 0.
* `kind/documentation` and `kind/bug`: if `kind/breaking` and `kind/feature` are
not present, next release is a PATCH version.

A Pull Request requires at least one of these labels.

== Release Notes

The KubeArchive release process creates release notes for users. These notes are
generated from the GitHub Pull Requests that contain a code block annotated with
`release-note`:

[source,text]
----
```release-note
Introduced new feature called `feature-a`. To enable it, set to `true` the
`feature-a` flag in the `feature-flags` ConfigMap in the `kubearchive` namespace.
```
----

To not include release notes from the Pull Request the release-note block must
contain the string `NONE`:

[source,text]
----
```release-note
NONE
```

> ----

Reviewer 2: Can we add an example of when this is expected to happen? I guess that adding tests would be an example of not needing release notes in a PR, right?

Author: I added some examples in the previous paragraphs. Let me know what you think

contain the string `NONE`:

[source,text]
----
```release-note
NONE
```
----

== Steps

. Merge the changes needed into `main`.
. Make sure the latest commit in `main` has passed every test.
. Go to the Actions tab on the KubeArchive repository.
. Trigger the "Release" workflow.

== Release Results

. A release commit is created adding the `VERSION` and `charts/kubearchive/Chart.yaml` files.
. A release tag is created from the release commit.
. A GitHub release is created from the release tag.
. Images are pushed with the `latest` and `<version>` tags to https://quay.io/kubearchive.
. The Helm chart is pushed to https://quay.io/kubearchive.

== Testing

The release workflow is prepared to be run within forks. First you need to set the
following variables in your fork:

* `OCI_REPOSITORY`: string exported as `KO_DOCKER_REPO` to upload images and used
in the `helm push` step as `oci://${OCI_REPOSITORY}`.

And the following secrets:

* `OCI_USERNAME`: username to authenticate with `ko login` and `helm registry login`
* `OCI_PASSWORD`: password to authenticate with `ko login` and `helm registry login`

To revert the effects of testing:

> . Delete the tag and release it creates from the GitHub web interface.

Reviewer 3: "Delete both the tag and release it created ..." WDYT?

Author: I rewrote the section, please check again.

= Release Process

> This document explains the requirements for and the process of a KubeArchive release.

Author: Its valid english, at least in the eyes of Chris. I changed it to a more friendly phrase.

> This document explains the requirements for and the process of a KubeArchive release.

Reviewer 2: oh, then that's my bad, but thanks for rephrasing it !

contain the string `NONE`:

[source,text]
----
```release-note
NONE
```
----

This code block is included in KubeArchive's Pull Request template.

== Steps

. Merge the changes needed into `main`.
. Make sure the latest commit in `main` has passed every test.
. Go to the Actions tab on the KubeArchive repository.
. Trigger the "Release" workflow.

== Release Results

. A release commit is created adding the `VERSION` files.
. A release tag is created from the release commit.
. A GitHub release is created from the release tag.
. Images are pushed with the `<version>` tag to https://quay.io/kubearchive.
. The Helm chart is pushed to https://quay.io/kubearchive (version is set by a
command line option).

== Testing

The release workflow is prepared to be run within forks. You can run it locally
or on your fork Actions.

=== Testing Locally

Read the top of the script and install the tools listed there. Export
the "externally provided variables" listed on the script with proper values.
These variables provide authentication for some of the tools, login manually with
the rest of the tools. Then run the script:

[source,bash]
----
bash hack/release.sh
----

> To cleanup the results of the release, complete the following stpes:

Reviewer 3: typo: steps

Author: Fixed
